CoreAnalyticSchema/src/au/gov/asd/tac/constellation/graph/schema/analytic/compatibility/AnalyticSchemaV4UpdateProvider.java

```diff
@@ -0,0 +1,88 @@
+import au.gov.asd.tac.constellation.graph.schema.SchemaFactoryUtilities;
+import au.gov.asd.tac.constellation.graph.schema.analytic.AnalyticSchemaFactory;
+import au.gov.asd.tac.constellation.graph.schema.analytic.concept.AnalyticConcept;
+import au.gov.asd.tac.constellation.graph.schema.type.SchemaVertexType;
+import au.gov.asd.tac.constellation.graph.versioning.SchemaUpdateProvider;
+import au.gov.asd.tac.constellation.graph.versioning.UpdateProvider;
+import java.util.HashMap;
+import java.util.Map;
+import org.openide.util.lookup.ServiceProvider;
+
+/**
+ * Upgrade Vertex Types that have changed.
+ * <p>
+ * The detection regex and validation regex was reviewed and improved.
+ *
+ * @author arcturus
+ */
+@ServiceProvider(service = UpdateProvider.class)
+public class AnalyticSchemaV4UpdateProvider extends SchemaUpdateProvider {
+
+    public static final int SCHEMA_VERSION_THIS_UPDATE = 4;
+
+    @Override
+    protected SchemaFactory getSchema() {
+        return SchemaFactoryUtilities.getSchemaFactory(AnalyticSchemaFactory.ANALYTIC_SCHEMA_ID);
+    }
+
+    @Override
+    public int getFromVersionNumber() {
+        return AnalyticSchemaV3UpdateProvider.SCHEMA_VERSION_THIS_UPDATE;
+    }
+
+    @Override
+    public int getToVersionNumber() {
+        return SCHEMA_VERSION_THIS_UPDATE;
+    }
+
+    @Override
+    protected void schemaUpdate(final StoreGraph graph) {
+        final int typeAttribute = AnalyticConcept.VertexAttribute.TYPE.get(graph);
+
+        final Map<String, SchemaVertexType> typesToUpgrade = new HashMap<>() {
+            {
+                put(AnalyticConcept.VertexType.MD5.getName(), AnalyticConcept.VertexType.MD5);
+                put(AnalyticConcept.VertexType.SHA1.getName(), AnalyticConcept.VertexType.SHA1);
+                put(AnalyticConcept.VertexType.SHA256.getName(), AnalyticConcept.VertexType.SHA256);
+                put(AnalyticConcept.VertexType.COUNTRY.getName(), AnalyticConcept.VertexType.COUNTRY);
+                put(AnalyticConcept.VertexType.GEOHASH.getName(), AnalyticConcept.VertexType.GEOHASH);
+                put(AnalyticConcept.VertexType.MGRS.getName(), AnalyticConcept.VertexType.MGRS);
+                put(AnalyticConcept.VertexType.MD5.getName(), AnalyticConcept.VertexType.MD5);
+                put(AnalyticConcept.VertexType.IPV4.getName(), AnalyticConcept.VertexType.IPV4);
+                put(AnalyticConcept.VertexType.IPV6.getName(), AnalyticConcept.VertexType.IPV6);
+                put(AnalyticConcept.VertexType.EMAIL_ADDRESS.getName(), AnalyticConcept.VertexType.EMAIL_ADDRESS);
+                put(AnalyticConcept.VertexType.HOST_NAME.getName(), AnalyticConcept.VertexType.HOST_NAME);
+                put(AnalyticConcept.VertexType.URL.getName(), AnalyticConcept.VertexType.URL);
+                put(AnalyticConcept.VertexType.TELEPHONE_IDENTIFIER.getName(), AnalyticConcept.VertexType.TELEPHONE_IDENTIFIER);
+            }
```

Comment:
Sonar is identifying using double brace initialisation as a bug. Consequently, it would be good to remove the need to use it

Reply:
Yep, easily fixed.